src/Data/VectorSpace/Core.agda

```diff
@@ -0,0 +1,94 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Abstract vector spaces.
+--
+-- A "vector space" is a Module with a commutative, homomorphic inner
+-- product and a complete set of building blocks for mapping the space.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.VectorSpace.Core where
+
+import Algebra.Module.Morphism.Structures as MorphismStructures
+
+open import Algebra        using (CommutativeRing)
+open import Algebra.Module using (Module)
+open import Algebra.Module.Construct.TensorUnit using (⟨module⟩)
+open import Algebra.Module.Morphism.Bundles
+open import Data.List      using (List; foldr)
+open import Data.Product
+open import Function
+open import Level          using (Level; _⊔_; suc)
+open import Relation.Binary
+open import Relation.Binary.ExtensionalEquivalence
+open import Relation.Binary.Reasoning.MultiSetoid
+
+private
+  variable
+    a b c : Level
+    A : Set a
+    B : Set b
+    C : Set c
+
+record VectorSpace
+  {r ℓr m ℓm : Level}
+  {ring      : CommutativeRing r ℓr}
+  (mod       : Module ring m ℓm)
+  : Set (suc (ℓr ⊔ r ⊔ ℓm ⊔ m)) where
+
+  constructor mkVS
+
+  open CommutativeRing ring renaming (Carrier  to S)   public
+  open Module          mod  renaming (Carrierᴹ to V)   public
+  open MorphismStructures.ModuleMorphisms mod ⟨module⟩ public
+
+  vscale : (V → S) → V → V
+  vscale f = uncurry _*ₗ_ ∘ < f , id >
+
+  vgen : (V → S) → List V → V
+  vgen f = foldr (_+ᴹ_ ∘ vscale f) 0ᴹ
+
+  infix 7 _∙_
+  field
+    _∙_           : V → V → S
+    basisSet      : List V
+    basisComplete : ∀ {a : V} →
+                    a ≈ᴹ vgen (a ∙_) basisSet
+
+    -- ToDo: Can these be unified, by using one of the
+    -- existing algebraic structures?
+    -- I'm only finding things that are predicated upon: `A → A → A`, or
+    -- `A → B`; nothing for: `A → A → B`.
```

Comment:
See `Algebra.Module.Definitions`

Reply:
Ah sorry, yes, the definitions I was thinking of are here:
https://github.com/agda/agda-stdlib/blob/d5ecb051ffb1965b2db45d09e330a13b3bf64471/src/Algebra/Module/Core.agda#L18-L22

But actually `A -> A -> B` is a bit weird as you're conjuring out an entirely new `B` from 2 `A`s. I'm not aware of any general name for such an operation, but you probably know the literature better than I do? I'd just call that a function... and hence not give it a name.

Reply:
Okay, thanks!